### PR TITLE
Skip rules with unsupported rule actions when matching requests

### PR DESCRIPTION
**Pull request by kzar**
With v4 of the block list, rules can have custom actions for features
such as Click to Load. Two changes are required:
 - Rules with unsupported actions should be skipped when matching
   requests.
 - Caller needs to be able to supply its own set of supported custom
   rule actions. That way, for example, if Facebook Click to Load is
   enabled for a page, the extension can also look for matching rules
   with the "block-ctl-fb" rule action.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,10 +19,9 @@
   "devDependencies": {
     "chalk": "2.4.0",
     "commander": "2.15.1",
-    "eslint": "^7.32.0",
-    "eslint-config-standard": "^16.0.3",
+    "eslint": "^8.31.0",
```
Comment (kzar) [inline package.json:22]: Required for support of the static class property syntax.

File: .eslintrc [hunk 2]
```diff
@@ -1,7 +1,8 @@
 {
     "extends": "standard",
     "rules": {
-        "indent": ["error", 4]
+        "indent": ["error", 4],
+        "object-shorthand": 0
```
Comment (kzar) [inline .eslintrc:5]: Required since we updating eslint required updating the standard config and the standard config started flagging these. I wanted to avoid making a bunch of unrelated changes to fix those for now.

File: src/classes/trackers.js [hunk 3]
```diff
@@ -78,6 +78,10 @@
  */
 
 class Trackers {
+    static supportedRuleActions = new Set([
+        'ad-attribution', 'block', 'ignore', 'ignore-user', 'redirect'
```
Comment (jdorweiler) [inline src/classes/trackers.js:82]: Where do ad-attribution and ignore-user come from?

Comment (kzar) [reply src/classes/trackers.js:82]: I think 'ignore-user' action is one that the extension code sets itself, but wouldn't be found in the block list. I considered not including it here, we might not need it. It seemed less risky to include it though in case there's a code-path I missed - perhaps the extension mutates the block list to change the rule action to 'ignore-user' sometimes?

I included 'ad-attribution' as it's included in the `ActionName` type above. (See https://github.com/duckduckgo/privacy-grade/blob/main/src/classes/trackers.js#L70 .) I don't know much about that one, but I think @jonathanKingston would know more.

Comment (jdorweiler) [reply src/classes/trackers.js:82]: This class is still this in MV3 to know what trackers were blocked by DNR rules? If so then I think we would show CTL requests as not being blocked.

Comment (kzar) [reply src/classes/trackers.js:82]: Yea, good point. It's possible the MV2 code-path would report these CTL requests as being allowed for MV3 before we eventually add MV2 support for the new CTL rule actions. We might have to just accept that, but I will mark this as a draft for now while we think about the trade-offs.

Comment (kzar) [reply src/classes/trackers.js]: > Where do ad-attribution and ignore-user come from? 

So I looked at the logic more today. I don't think we need these other actions, they aren't in the block list but are instead returned in the tracker Object (by `getTrackerData`). For example, if the tracker rule has a surrogate property, the action is reported as 'redirect' while the rule itself is left alone.

> This class is still this in MV3 to know what trackers were blocked by DNR rules? If so then I think we would show CTL requests as not being blocked.

Once both the MV2 and MV3 extension are using these Click to Load rule actions for blocking/redirecting Click to Load requests we should be good. They should be reported as well for MV3 as other blocked/redirected requests, since the `onBeforeRequest` code-path is still being run on MV3 for now.

**Review by ladamski — CHANGES_REQUESTED** [review]

File: src/classes/trackers.js [hunk 4]
```diff
@@ -452,18 +462,38 @@ class Trackers {
...
      * @returns {TrackerRule | null}
      */
-    findRule (tracker, requestData) {
+    findRule (tracker, requestData, supportedCustomRuleActions) {
         let matchedRule = null
         // Find a matching rule from this tracker
         if (tracker.rules && tracker.rules.length) {
             tracker.rules.some(ruleObj => {
-                if (this.requestMatchesRule(requestData, ruleObj)) {
+                if (this.ruleActionSupported(ruleObj, supportedCustomRuleActions) &&
```
[the remark this answers is not in the document]
Comment (kzar) [reply src/classes/trackers.js:495]: > 1. Recommendation: for performance reasons...

Interesting. I actually put the checks this way around for performance reasons. My thinking was that a Set lookup (or even two) for the rule action is much cheaper an operation than checking if the rule matches (checking if a regular expression matches the URL, plus more work if the rule has extra options.) But you're right, more often than not the rule action _will_ be supported.

Which is faster overall, avoiding checking if the rules match requests for some inactive Click to Load rules, or skipping the rule action Set lookups for rules that don't match a request... I'm not sure! My vote would be to leave this as is for now, but to revisit if performance turns out to be an issue WDYT?

> 2. Idea: another approach would be to just filter out unrecognized rule actions when parsing the rule set?

I don't think this is possible unfortunately. Supposing Facebook Click to Load is disabled for a tab, we _don't_ want to support the "block-ctl-fb" rule action any more for the tab's requests. I think we need the ability to specify which rule actions are active each time we check for matches.

[the remark this answers is not in the document]
Comment (kzar) [reply src/classes/trackers.js]: > 1. Recommendation: for performance reasons...

As discussed in MatterMost, I was on the fence about which was better for performance in practice without measuring it. Since you felt more strongly, I just swapped it. Done.

**Review by jdorweiler — APPROVED** [review]
